File: spherepop-kernel/src/arbiter.rs
//! The Arbiter: the only path by which a `History` is ever extended.
//!
//! `validate` reads `State` — which carries only `(H, Omega)`-level facts —
//! and never the value any collapse rule would compute. There is no
//! parameter through which `c(H)` could be passed in, for any `c`: this is
//! Requirement (Observation Non-Interference) enforced by the type
//! signature, not by convention.

use crate::event::{Event, EventKind, LogPos, ObjectId, RuleId};
use crate::history::{History, State};
use std::collections::HashSet;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ArbiterError {
    Malformed(String),
    PopOutsideOptionSpace(ObjectId),
    UncertifiedCollapseRule(RuleId),
    RefuseWithoutReason,
    StaleOverlay,
}

impl std::fmt::Display for ArbiterError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}
impl std::error::Error for ArbiterError {}

pub struct Proposal {
    pub events: Vec<Event>,
}

impl Proposal {
    pub fn new(events: Vec<Event>) -> Self {
        Proposal { events }
    }
}

pub struct Arbiter {
    history: History,
    /// The admissibility-certified rule registry: an uncertified rule
    /// name is a type error at proposal time, matching the primitive
    /// Collapse definition's precondition Gamma |- t : Adm(T).
    rules: HashSet<RuleId>,
    omega_0: HashSet<ObjectId>,
}

impl Arbiter {
    pub fn new(omega_0: impl IntoIterator<Item = ObjectId>, rules: impl IntoIterator<Item = RuleId>) -> Self {
        Arbiter {
            history: History::new(),
            rules: rules.into_iter().collect(),
            omega_0: omega_0.into_iter().collect(),
        }
    }

    pub fn len(&self) -> usize {
        self.history.len()
    }

    pub fn history_ref(&self) -> &History {
        &self.history
    }

    pub fn history_clone(&self) -> History {
        self.history.clone()
    }

    pub fn state(&self) -> State {
        self.history.replay(&self.omega_0)
    }

    pub fn omega_0_ref(&self) -> &HashSet<ObjectId> {
        &self.omega_0
    }

    /// Validates and, if accepted, appends a proposal to H. The ONLY path
    /// by which H is ever extended.
    pub fn submit(&mut self, p: Proposal) -> Result<Vec<LogPos>, ArbiterError> {
        self.validate(&p.events)?;
        let mut positions = Vec::new();
        for mut e in p.events {
            let pos = self.history.len() as LogPos;
            e.pos = pos;
            self.history.push(e);
            positions.push(pos);
        }
        Ok(positions)
    }

    fn validate(&self, events: &[Event]) -> Result<(), ArbiterError> {
        // Structural state only: (H, Omega)-level facts, never c(H) for any c.
        let s = self.history.replay(&self.omega_0);
        let mut hypothetical_committed: HashSet<ObjectId> = HashSet::new();

        for e in events {
            match e.kind {
                EventKind::Pop => {
                    let x = e.a.ok_or_else(|| ArbiterError::Malformed("Pop missing a".into()))?;
                    let still_available = s.option_space.contains(&x) && !hypothetical_committed.contains(&x);
                    if !still_available {
                        return Err(ArbiterError::PopOutsideOptionSpace(x));
                    }
                    hypothetical_committed.insert(x);
                }
                EventKind::Refuse => {
                    if e.reason.as_deref().unwrap_or("").is_empty() {
                        return Err(ArbiterError::RefuseWithoutReason);
                    }
                }
                EventKind::Bind => {
                    if e.a.is_none() || e.b.is_none() {
                        return Err(ArbiterError::Malformed("Bind missing a/b".into()));
                    }
                }
                EventKind::Collapse => {
                    let rule = e.rule.ok_or_else(|| ArbiterError::Malformed("Collapse missing rule".into()))?;
                    if !self.rules.contains(rule) {
                        return Err(ArbiterError::UncertifiedCollapseRule(rule));
                    }
                }
            }
        }
        Ok(())
    }
}

File: spherepop-go/arbiter.go
package spherepop

import "fmt"

type ErrorCode string

const (
	Malformed               ErrorCode = "Malformed"
	PopOutsideOptionSpace   ErrorCode = "PopOutsideOptionSpace"
	UncertifiedCollapseRule ErrorCode = "UncertifiedCollapseRule"
	RefuseWithoutReason     ErrorCode = "RefuseWithoutReason"
	StaleOverlay            ErrorCode = "StaleOverlay"
)

type ArbiterError struct {
	Code   ErrorCode
	Detail string
}

func (e *ArbiterError) Error() string {
	if e.Detail == "" {
		return string(e.Code)
	}
	return fmt.Sprintf("%s(%s)", e.Code, e.Detail)
}

type Proposal struct{ Events []Event }

func NewProposal(events ...Event) Proposal {
	return Proposal{Events: append([]Event(nil), events...)}
}

type Arbiter struct {
	history History
	rules   map[RuleID]struct{}
	omega0  []ObjectID
}

func NewArbiter(omega []ObjectID, certifiedRules []RuleID) *Arbiter {
	a := &Arbiter{
		history: NewHistory(),
		rules:   make(map[RuleID]struct{}, len(certifiedRules)),
		omega0:  append([]ObjectID(nil), omega...),
	}
	for _, id := range certifiedRules {
		a.rules[id] = struct{}{}
	}
	return a
}

func (a *Arbiter) Len() int         { return a.history.Len() }
func (a *Arbiter) History() History { return History{events: a.history.Events()} }
func (a *Arbiter) State() State     { return a.history.Replay(a.omega0) }

// Submit atomically validates the complete proposal against the state after
// each preceding proposed event, then appends every event or none of them.
func (a *Arbiter) Submit(p Proposal) ([]LogPos, error) {
	if err := a.validate(p.Events); err != nil {
		return nil, err
	}
	positions := make([]LogPos, 0, len(p.Events))
	for _, event := range p.Events {
		event.Pos = LogPos(a.history.Len())
		a.history.append(event)
		positions = append(positions, event.Pos)
	}
	return positions, nil
}

func (a *Arbiter) validate(events []Event) error {
	hypothetical := a.State()
	for _, event := range events {
		switch event.Kind {
		case Pop:
			if event.A == nil {
				return &ArbiterError{Malformed, "Pop missing a"}
			}
			if _, ok := hypothetical.OptionSpace[*event.A]; !ok {
				return &ArbiterError{PopOutsideOptionSpace, fmt.Sprint(*event.A)}
			}
		case Refuse:
			if event.Reason == nil || *event.Reason == "" {
				return &ArbiterError{RefuseWithoutReason, ""}
			}
		case Bind:
			if event.A == nil || event.B == nil {
				return &ArbiterError{Malformed, "Bind missing a/b"}
			}
		case Collapse:
			if event.Rule == nil {
				return &ArbiterError{Malformed, "Collapse missing rule"}
			}
			if _, ok := a.rules[*event.Rule]; !ok {
				return &ArbiterError{UncertifiedCollapseRule, string(*event.Rule)}
			}
		default:
			return &ArbiterError{Malformed, "unknown event kind"}
		}
		Apply(&hypothetical, event)
	}
	return nil
}
